numba/cpython/unicode.py

```diff
@@ -842,6 +830,56 @@ def _adjust_indices(length, start, end):
     return start, end
 
 
+@overload_method(types.UnicodeType, 'startswith')
+def unicode_startswith(s, prefix, start=None, end=None):
+    if not (start is None or isinstance(start, (types.Omitted,
+                                                types.Integer,
+                                                types.NoneType))):
+        raise TypingError('When specified, the arg "start" must be an Integer')
+
+    if not (end is None or isinstance(end, (types.Omitted,
+                                            types.Integer,
+                                            types.NoneType))):
+        raise TypingError('When specified, the arg "end" must be an Integer')
+
+    if isinstance(prefix, (types.Tuple, types.UniTuple)):
```

Comment:
`Tuple` cannot be accepted here unless the contents of the `Tuple` are all `types.StringLiteral`. Consider:

```python
@njit
def foo():
    print('abcde'.startswith(('a', 1)))

foo()
```
it may be easier to just not accept `Tuple` to force type inference to try with a non-literal types instead.

Reply:
I prefer to accept `Tuple` of `StringLiteral`, 1) it's more consistent with cpython, 2) I do need this to simplify my next PR https://github.com/numba/numba/pull/8558. So I prefer to add a check to make sure every item in `Tuple` (or `UniTuple`) is a `StringLiteral`.

Reply:
Revised, please review this issue again.

Reply:
By default, type inference will try using non-literal types first, and then if that fails, it will try again using literal types. Python source containing tuples of string literals will be inferred as `types.Tuple(<types.StringLiteral>, <types.StringLiteral>, ...)`. e.g.
```python
from numba import njit
import numpy as np

@njit
def foo():
    x = ('a', 'b')

foo()
foo.inspect_types()
```
-------------------------------------
You can see the "try non-literal then literal" types behaviour by doing something like:
```python
from numba import njit, types
from numba.extending import overload
import numpy as np

def bar(tup):
    pass

@overload(bar)
def ol_bar(tup):
    print(f"bar:: tup type is {tup}")

@njit
def foo():
    x = ('a', 'b')
    bar(x)

foo()
```
(ignore the compilation error, just look at what gets printed to stdout).

-------------------------------------
RE: the `'a'.startswith(('a', 5))`, there are implementation details like this in the CPython implementation. They don't need replicating unless there's some proven use case!